File: practise/leetcode/DailyQuestion/202406/3040.py
class Solution:
    def maxOperations(self, nums: List[int]) -> int:
        @cache
        def dfs(i: int, j: int, target: int) -> int:
            if i >= j:
                return 0;
            res = 0
            if nums[i] + nums[i +1 ] == target:
                res = max(res, dfs(i + 2, j, target) + 1)
            if nums[j - 1] + nums[j] == target:
                res = max(res, dfs(i, j - 2,target) + 1)
            if nums[i] + nums[j] == target:
                res = max(res, dfs(i + 1, j - 1, target) + 1)
            return res

        n = len(nums)
        res1 = dfs(2, n - 1, nums[0] + nums[1])
        res2 = dfs(0, n - 3, nums[-2] + nums[-1])
        res3 = dfs(1, n - 2, nums[0] + nums[-1])
        return max(res1, res2, res3) + 1

File: practise/leetcode/DailyQuestion/202406/3040.cpp
class Solution {
public:
    int maxOperations(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(n));
        auto solve = [&](int i,int j,int target) -> int{
            for(auto & row : dp){
                ranges::fill(row, -1);
            }
            function<int(int,int)> dfs = [&](int i,int j) -> int{
                if(i >= j) return 0;
                int &res=dp[i][j];
                if (res != -1) return res;
                res = 0;
                if(nums[i]+nums[i+1]==target) res = max(res, dfs(i + 2, j) + 1);
                if(nums[j-1]+nums[j]==target) res = max(res, dfs(i, j - 2) + 1);
                if(nums[i]+nums[j]==target) res = max(res, dfs(i + 1, j - 1) + 1);
                return res;
            }; 
            return dfs(i, j);
        };
        int res1 = solve(2, n - 1, nums[0] + nums[1]);
        int res2 = solve(0, n - 3, nums[n - 2] + nums[n - 1]);
        int res3 = solve(1, n - 2, nums[0] + nums[n - 1]);
        return max({res1,res2,res3}) + 1;
    }
};